[ARM]Adjust cost of muls in (U/S)MLAL and patterns

nasherm opened the pull request:
PR #117350 made changes to the SLP vectorizer which introduced a regression on some ARM  benchmarks. Investigation narrowed it down to suboptimal codegen for benchmarks that previously only used scalar SMLAL instructions. The linked change meant the SLPVectorizer thought that these could be vectorized. This change makes the cost of muls in (U/S)MLAL patterns slightly cheaper to make sure scalar instructions are preferred in these cases over SLP vectorization

nasherm committed:
[ARM][SLP] Fix incorrect cost function for SLP Vectorization of ZExt/SExt

PR #117350 made changes to the SLP vectorizer which introduced
a regression on ARM vectorization benchmarks. This was due
to the changes assuming that SExt/ZExt vector instructions have
constant cost. This behaviour is expected for RISCV but not on ARM
where we take into account source and destination type of SExt/ZExt
instructions when calculating vector cost.

Change-Id: I6f995dcde26e5aaf62b779b63e52988fb333f941

davemgreen commented on llvm/test/Transforms/SLPVectorizer/ARM/expensive-arithmetic-extended-reduction-mve.ll:
```diff
@@ -0,0 +1,354 @@
+; CHECK-NEXT:    [[TMP21:%.*]] = add nsw i64 [[TMP16]], [[TMP20]]
+; CHECK-NEXT:    ret i64 [[TMP21]]
+;
+  %2 = alloca ptr, align 4
```
Can you run these tests through opt -O1 (or sroa/mem2reg) to clean up the allocas. They might be confusing the SLP vectorizer on their own.

nasherm replied:
Done

davemgreen reviewed (commented):
The MVE instruction that getExtendedReductionCost is trying to model is vaddlv, which is a `i64 vecreduce(v4i64 sext(v4i32))`. So it would seem odd to add the cost of the extend again to it.  BaseT::getExtendedReductionCost includes the cast cost too, so it would look like this is double-counting.

This is an example of MVE using the vaddlv instruction, which looks decent in this small example: https://llvm.godbolt.org/z/oTrMTcM4E

nasherm commented:
My most recent patch fixes the regression with a one line change to ARMTTI. From what I understand, MVE vectors greater than 128 bits aren't handled very well. It appears that the changes to SLP revealed a bug where we were allowing extensions of the type ` vadd v4i32 -> vadd v4i64`. 

@davemgreen does my fix make sense with this reasoning? I'm skeptical as I suspect I'm missing something which justifies the original code which allowed this extension.

davemgreen reviewed (commented):
It is usually the input type that matters for MVE, and a vadd v4i32 -> vadd v4i64 reduction is still a single vaddlv. The comment IIRC was talking about things like v8i32->i64 reductions, that would need to be split into two v4i32 reductions but is difficult under MVE due to the way predicates and extensions work.

There are some Loop-vectorize test that would need to be updated, but I believe we would still want to generate an extending reduction pattern. There might be something more complex going on where a mixture of MVE and scalar code doesn't perform as well as sticking with scalar. Or maybe the SLP vectorizer isn't handling multiple uses properly, or the scalar cost should be a little lower for SMLAL DSP instructions.

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/ARM/vadd-mve.s:
```diff
@@ -0,0 +1,161 @@
+	.text
+	.syntax unified
```
Not a test, delete

nasherm replied:
Done

alexey-bataev commented on llvm/test/Transforms/SLPVectorizer/ARM/vadd-mve.ll:
```diff
@@ -0,0 +1,222 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes="default<O1>,slp-vectorizer" --mtriple arm-none-eabi -mattr=+mve -S -o - | FileCheck %s
```
O1 should not be used here, only slp-vectorizer

nasherm replied:
Done

nasherm committed:
Adjust extended reduction cost

Change-Id: Ie2795e0e5ebb0589146eaf07c752410e307a36e6

nasherm commented:
I've opted for adjusting the cost returned by getExtendedReductionCost for vadd reductions to be slightly more pessimistic. This returns performance numbers on regressed benchmarks while not changing SLPVectorizer and making minimal changes to the rest of the compiler.

davemgreen commented:
There is a test failing, which I think says "we shouldn't do this". I would expect the cost of a VADDV and VADDLV to be about the same as other MVE instructions, unless we have a strong reason otherwise.

nasherm committed:
[ARM]Reduce scalar cost of SMLAL muls

Changes to the SLPVectorizer caused a regression
in some benchmarks which targetted cores that support
both DSP and MVE instructions. The particular regression
has been reduced to a case where MUL instructions that
are part of a chain of instructions that can be replaced
with DSP SMLAL may also be vectorized. The generated
code ends up being an inefficient of both scalar
and vector ops rather than leaning to one or the other.
By reducing the cost of these MUL instructions in these
patterns we recover lost performance.

Change-Id: I302817cf4fcd18a11d40fba430c44e034a36448b

nasherm commented:
Update:
After revisiting this I narrowed down the cause of our regressions. Changes to the SLPVectorizer caused a regression
in some benchmarks which targeted cores that support
both DSP and MVE instructions. The particular regression
has been reduced to a case where MUL instructions that
are part of a chain of instructions that can be replaced
with DSP SMLAL may also be vectorized. The generated
code ends up being an inefficient combination of both scalar
and vector ops rather than leaning to one or the other.
By reducing the cost of these MUL instructions in these
patterns we recover lost performance.

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+    auto *BinOp = dyn_cast<BinaryOperator>(I);
+    if (!BinOp)
+      return false;
+    auto *Op0 = BinOp->getOperand(0);
```
auto -> Value

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      return false;
+
+    auto IsSExtInst = [](const Value *V) -> bool {
+      return (dyn_cast<SExtInst>(V)) ? true : false;
```
return isa<SExtInst>(V);

Is it worth adding ZExt at the same time for UMULL?

nasherm replied:
Done. Surprisingly no. Vectorization for my reproducer using unsigned types is done quite well

nasherm replied:
So by adding ZExt we do avoid mixing vector and scalar instructions. The generated code doesn't take advantage of UMULL instructions however and just sticks with simple scalar ops. This was the behavior before the changes to the SLPVectorizer were made.

I do think that once this is merged there is further improvement to be done on SMLAL code gen which I intend to investigate i.e  further code folding using DSP instructions. I could add UMULL codegen into my investigation as well. 

Regardless adding ZExt support will just return previous behavior so I don't believe it to be a concern.

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      return false;
+    auto *Op0 = BinOp->getOperand(0);
+    auto *Op1 = BinOp->getOperand(1);
+    if (Op0 && Op1 && IsSExtInst(Op0) && IsSExtInst(Op1)) {
```
Don't need to check Op0 && Op1 here - they will always be non-null.

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+    auto *Op1 = BinOp->getOperand(1);
+    if (Op0 && Op1 && IsSExtInst(Op0) && IsSExtInst(Op1)) {
+      // In this case we're interested in an ext of an i16
+      if (!Op0->getType()->isIntegerTy(32) || !Op1->getType()->isIntegerTy(32))
```
Maybe check Op0->getOperand(0)->getType() is an i16, Op1->getOperand(0)->getType() is a i16 and I->getType() (which is the same as Op0->getType()) is an i32?

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      if (!Op0->getType()->isIntegerTy(32) || !Op1->getType()->isIntegerTy(32))
+        return false;
+      // We need to check if this result will be further extended to i64
+      for (auto *U : I->users())
+        if (IsSExtInst(dyn_cast<Value>(U)))
+          return true;
```
Probably check oneuse and it isa SExt, or all uses are sext's?

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,62 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
   if (LooksLikeAFreeShift())
     return 0;
 
+  // When targets have both DSP and MVE we find that the
```
Is it worth reducing the cost when we don't have MVE (but do have DSP) too? LLVM isn't great at trying to cost-model multiple instructions that can combine into one.  I think I would expect the mul to still have a cost, but the sext's to be free - I'm not sure if that would do better or worse in practice.

nasherm replied:
Sure. Seems to make sense

davemgreen commented on llvm/test/Analysis/CostModel/ARM/muls-in-smlal-patterns.ll:
```diff
@@ -0,0 +1,37 @@
+; RUN: opt -passes="print<cost-model>" 2>&1 -disable-output -mtriple thumbv8.1-m.main -mattr=+mve,+dsp  < %s | FileCheck %s
+define i64 @test(i16 %a, i16 %b) {
```
Can you update with update_analysis_test_checks? And maybe add a run line without +dsp to show the differences.

davemgreen reviewed (commented):
Sounds good - it looks like this will work quite well.

nasherm committed:
Responding to review comments

Change-Id: Ie8c40972ae07e568d767ace37b9dda0f77272569

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,76 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+    };
+    auto IsExtensionFromHalf = [&, IsSExtInst,
+                                IsZExtInst](const Value *V) -> bool {
+      if (IsSExtInst(V))
+        return dyn_cast<SExtInst>(V)->getOperand(0)->getType()->isIntegerTy(16);
+      if (IsZExtInst(V))
+        return dyn_cast<ZExtInst>(V)->getOperand(0)->getType()->isIntegerTy(16);
+      return false;
```
`return cast<Instruction>(V)->getOperand(0)->getType()...`

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,76 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+    if (Ty->isVectorTy())
+      return false;
+
+    auto IsSExtInst = [](const Value *V) -> bool { return isa<SExtInst>(V); };
+    auto IsZExtInst = [](const Value *V) -> bool { return isa<ZExtInst>(V); };
```
It might be simpler to drop these and test isa<..> directly instead.

nasherm replied:
Done

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,76 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      return false;
+    Value *Op0 = BinOp->getOperand(0);
+    Value *Op1 = BinOp->getOperand(1);
+    if (IsExtInst(Op0) && IsExtInst(Op1)) {
```
Should it check that both Ops have the same opcode?

nasherm replied:
Yes. That makes sense. Done. Also added extra test case

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,76 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      // We need to check if this result will be further extended to i64
+      // and that all these uses are SExt
+      for (auto *U : I->users())
+        if (!IsExtInst(dyn_cast<Value>(U)))
```
Does IsExtInst(U) work?

nasherm replied:
Yes, done

davemgreen commented on llvm/test/Analysis/CostModel/ARM/muls-in-smlal-patterns.ll:
```diff
@@ -0,0 +1,80 @@
+; CHECK-NO-DSP-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %ms = sext i32 %m to i64
+; CHECK-NO-DSP-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: ret i64 %ms
+;
+    %as = sext i16 %a to i32
+    %bs = sext i16 %b to i32
```
Add tests for zext too?

nasherm replied:
Done

davemgreen reviewed (commented):
Thanks - other than the inline comments this looks good to me.

nasherm committed:
Further review comments

Change-Id: I0c20291a8a1d2cbc895890d02aea963323ee48cb

davemgreen commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,75 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+      return isa<ZExtInst>(V) || isa<SExtInst>(V);
+    };
+    auto IsExtensionFromHalf = [&, IsExtInst](const Value *V) -> bool {
+      if (IsExtInst(V))
```
I think at this point we know that IsExtInst is true?

nasherm replied:
True :)

davemgreen reviewed (approved):
LGTM, thanks.

nasherm committed:
Minor comments

Change-Id: I9ce62da8405a08ece5e36c7e1ea0bb93e6f0b6c4

fhahn commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,73 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+    auto IsExtInst = [](const Value *V) -> bool {
+      return isa<ZExtInst>(V) || isa<SExtInst>(V);
+    };
+    auto IsExtensionFromHalf = [&, IsExtInst](const Value *V) -> bool {
```
`IsExtInst` not used?

nasherm replied:
Hi Florian, yes.  I saw it triggered some buildbot failures. I opened #132018 to fix this. Is it preferable to use this merged branch?

fhahn commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1458,16 +1458,73 @@ InstructionCost ARMTTIImpl::getArithmeticInstrCost(
+        return false;
+      // We need to check if this result will be further extended to i64
+      // and that all these uses are SExt
+      for (auto *U : I->users())
+        if (!IsExtInst(U))
+          return false;
+      return true;
```
can this just be

`return all_of(I->users(), IsExtInst)`?
